# Ingeniería de Características: Creación de Nuevas Variables

Cargamos los datos

In [351]:
import pandas as pd
ds_path = "data/train.csv"
df = pd.read_csv(ds_path)

In [352]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

Visualizamos los features y tipos de datos

In [353]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Analizamos las estadisticas descriptivas del df

In [354]:
from utils.eda import get_descriptive_statistics
get_descriptive_statistics(df).T

min       max        mean         std    median  \
PassengerId  1.00  891.0000  446.000000  257.353842  446.0000   
Survived     0.00    1.0000    0.383838    0.486592    0.0000   
Pclass       1.00    3.0000    2.308642    0.836071    3.0000   
Age          0.42   80.0000   29.699118   14.526497   28.0000   
SibSp        0.00    8.0000    0.523008    1.102743    0.0000   
Parch        0.00    6.0000    0.381594    0.806057    0.0000   
Fare         0.00  512.3292   32.204208   49.693429   14.4542   

             variation_coefficient  
PassengerId               0.577027  
Survived                  1.267701  
Pclass                    0.362149  
Age                       0.489122  
SibSp                     2.108464  
Parch                     2.112344  
Fare                      1.543073

Obtendremos el titulo de cada persona a partir del nombre

In [355]:
df["Name"]

0                                Braund, Mr. Owen Harris
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                               Allen, Mr. William Henry
                             ...                        
886                                Montvila, Rev. Juozas
887                         Graham, Miss. Margaret Edith
888             Johnston, Miss. Catherine Helen "Carrie"
889                                Behr, Mr. Karl Howell
890                                  Dooley, Mr. Patrick
Name: Name, Length: 891, dtype: object

Extraemos el titulo y visualizamos

In [356]:
df['Title'] = df['Name'].str.extract(r'([A-Za-z]+)\.', expand=True)
df["Title"].unique().tolist()

['Mr',
 'Mrs',
 'Miss',
 'Master',
 'Don',
 'Rev',
 'Dr',
 'Mme',
 'Ms',
 'Major',
 'Lady',
 'Sir',
 'Mlle',
 'Col',
 'Capt',
 'Countess',
 'Jonkheer']

Simplificamos los valores categoricos

In [357]:
def _convertir_titulo(X):
    titles = {
        "Mr": ["Mr", "Master", "Dr", "Col", "Capt", "Jonkheer", "Don", "Rev", "Sir", "Major"],
        "Mrs": ["Mrs", "Ms", "Mme", "Lady", "Mlle", "Countess"],
        "Miss": ["Miss"],
    }

    for key, values in titles.items():
        if X in values:
            return key

    return X


df["Title"] = df["Title"].apply(_convertir_titulo)
df["Title"].unique().tolist()

['Mr', 'Mrs', 'Miss']

Observamos los valores categoricos del feature Cabin

In [358]:
df["Cabin"].unique().tolist()

[nan,
 'C85',
 'C123',
 'E46',
 'G6',
 'C103',
 'D56',
 'A6',
 'C23 C25 C27',
 'B78',
 'D33',
 'B30',
 'C52',
 'B28',
 'C83',
 'F33',
 'F G73',
 'E31',
 'A5',
 'D10 D12',
 'D26',
 'C110',
 'B58 B60',
 'E101',
 'F E69',
 'D47',
 'B86',
 'F2',
 'C2',
 'E33',
 'B19',
 'A7',
 'C49',
 'F4',
 'A32',
 'B4',
 'B80',
 'A31',
 'D36',
 'D15',
 'C93',
 'C78',
 'D35',
 'C87',
 'B77',
 'E67',
 'B94',
 'C125',
 'C99',
 'C118',
 'D7',
 'A19',
 'B49',
 'D',
 'C22 C26',
 'C106',
 'C65',
 'E36',
 'C54',
 'B57 B59 B63 B66',
 'C7',
 'E34',
 'C32',
 'B18',
 'C124',
 'C91',
 'E40',
 'T',
 'C128',
 'D37',
 'B35',
 'E50',
 'C82',
 'B96 B98',
 'E10',
 'E44',
 'A34',
 'C104',
 'C111',
 'C92',
 'E38',
 'D21',
 'E12',
 'E63',
 'A14',
 'B37',
 'C30',
 'D20',
 'B79',
 'E25',
 'D46',
 'B73',
 'C95',
 'B38',
 'B39',
 'B22',
 'C86',
 'C70',
 'A16',
 'C101',
 'C68',
 'A10',
 'E68',
 'B41',
 'A20',
 'D19',
 'D50',
 'D9',
 'A23',
 'B50',
 'A26',
 'D48',
 'E58',
 'C126',
 'B71',
 'B51 B53 B55',
 'D49',
 'B5',
 'B20',
 'F G

Observamos valores nulos, procedemos a imputarlos con el valor desconocido

In [359]:
from utils.eda import get_nulll_data_info
df["Cabin"] = df["Cabin"].fillna("UNK")
get_nulll_data_info(df)

datos sin NAs en q  Na en q  Na en %
Age                         714      177    19.87
Embarked                    889        2     0.22
Pclass                      891        0     0.00
Survived                    891        0     0.00
PassengerId                 891        0     0.00
Sex                         891        0     0.00
Name                        891        0     0.00
Parch                       891        0     0.00
SibSp                       891        0     0.00
Ticket                      891        0     0.00
Fare                        891        0     0.00
Cabin                       891        0     0.00
Title                       891        0     0.00

Observamos valores nulos en la columna Embarked, procedemos a imputarlos

In [360]:
df["Embarked"] = df["Embarked"].fillna("UNK")
get_nulll_data_info(df)

datos sin NAs en q  Na en q  Na en %
Age                         714      177    19.87
Survived                    891        0     0.00
Pclass                      891        0     0.00
Name                        891        0     0.00
PassengerId                 891        0     0.00
Sex                         891        0     0.00
SibSp                       891        0     0.00
Parch                       891        0     0.00
Ticket                      891        0     0.00
Fare                        891        0     0.00
Cabin                       891        0     0.00
Embarked                    891        0     0.00
Title                       891        0     0.00

Observamos valores nulos en la columna Age, imputamos aplicando la edad mediana

In [361]:
df['Age'] = df['Age'].fillna(df['Age'].median())
get_nulll_data_info(df)

datos sin NAs en q  Na en q  Na en %
PassengerId                 891        0      0.0
Survived                    891        0      0.0
Pclass                      891        0      0.0
Name                        891        0      0.0
Sex                         891        0      0.0
Age                         891        0      0.0
SibSp                       891        0      0.0
Parch                       891        0      0.0
Ticket                      891        0      0.0
Fare                        891        0      0.0
Cabin                       891        0      0.0
Embarked                    891        0      0.0
Title                       891        0      0.0

Ya no quedan datos nulos

Ahora procedemos a convertir las columnas booleanas

In [362]:
potential_boolean_values = [
    [0, 1],
    [0.0, 1.0],
    ["false", "true"],
    ["no", "yes"],
    ["no", "si"],
    ["N", "Y"],
]

boolean_sets = [set(map(str, values_list)) for values_list in potential_boolean_values]

def is_boolean_column(unique_values):
    unique_values = set(map(str, unique_values))
    return any(unique_values == boolean_set for boolean_set in boolean_sets)

def convert_to_boolean(df, column):
    false_values = set(map(str, [0, 0.0, "false", "no", "N"]))
    df[column] = df[column].apply(lambda x: x if pd.isna(x) else str(x).lower() in false_values)

columns = df.columns.tolist()
converted_boolean_columns = []

for column in columns:
    unique_values = df[column].dropna().unique()
    if len(unique_values) == 2 and is_boolean_column(unique_values):
        print(f"Column '{column}' es booleana, lo convertimos a tipo booleana.")
        convert_to_boolean(df, column)
        df[column] = df[column].astype("bool")
        converted_boolean_columns.append(column)

print(f"Columna convertida a booleana: {converted_boolean_columns}")

Column 'Survived' es booleana, lo convertimos a tipo booleana.
Columna convertida a booleana: ['Survived']


In [363]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    bool   
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        891 non-null    object 
 11  Embarked     891 non-null    object 
 12  Title        891 non-null    object 
dtypes: bool(1), float64(2), int64(4), object(6)
memory usage: 84.5+ KB


In [364]:
potential_categorical_columns = df.select_dtypes(include=["object", "string"]).columns.tolist()

potential_categorical_columns

['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked', 'Title']

Eliminamos espacios y caracteres invalidos de las cadenas

In [365]:
from utils.eda import clean_string


for column in potential_categorical_columns:
    df[column] = df[column].apply(clean_string)

Convertimos las columnas a categorica o cadena de acuerdo al porcentaje de valores unicos en comparación al total

In [366]:
max_percentage_categorical = 0.05
num_rows_df = df.shape[0]
converted_to_categorical = []
converted_to_string = []
for column in potential_categorical_columns:
    unique_values_count = df[column].nunique(dropna=True)
    print(f"Cantidad de valores unicos en la columna '{column}': {unique_values_count}")
    if unique_values_count <= (max_percentage_categorical * num_rows_df):
        df[column] = df[column].astype("category")
        converted_to_categorical.append(column)
    else:
        df[column] = df[column].astype("string")
        converted_to_string.append(column)
print("\n")
print(f"Número de columnas analizadas: {len(potential_categorical_columns)}")
print(f"Número de columnas convertidas a category: {len(converted_to_categorical)}")
print(f"Número de columnas convertidas a string: {len(converted_to_string)}")
print(f"Columnas convertidas a category: {converted_to_categorical}")
print(f"Columnas convertidas a string: {converted_to_string}")



Cantidad de valores unicos en la columna 'Name': 891
Cantidad de valores unicos en la columna 'Sex': 2
Cantidad de valores unicos en la columna 'Ticket': 681
Cantidad de valores unicos en la columna 'Cabin': 148
Cantidad de valores unicos en la columna 'Embarked': 4
Cantidad de valores unicos en la columna 'Title': 3


Número de columnas analizadas: 6
Número de columnas convertidas a category: 3
Número de columnas convertidas a string: 3
Columnas convertidas a category: ['Sex', 'Embarked', 'Title']
Columnas convertidas a string: ['Name', 'Ticket', 'Cabin']


Eliminamos las columnas que no utilizaremos y que no aportan información importante

In [367]:
df = df.drop("PassengerId", axis=1)
df = df.drop("Name", axis=1)
df = df.drop("Ticket", axis=1)

Creamos la característica FamilySize

In [368]:
df['FamilySize'] = df['SibSp'] + df['Parch'] + 1 # el 1 incluye al pasajero

In [369]:
df["FamilySize"]

0      2
1      2
2      1
3      2
4      1
      ..
886    1
887    1
888    4
889    1
890    1
Name: FamilySize, Length: 891, dtype: int64

In [370]:
def _convert_unknown_cabin(X):
    if X == "unk":
        return True

    return False


df["CabinUnknown"] = df["Cabin"].apply(_convert_unknown_cabin)

In [371]:
get_descriptive_statistics(df).T    

min       max       mean        std   median  \
Pclass      1.00    3.0000   2.308642   0.836071   3.0000   
Age         0.42   80.0000  29.361582  13.019697  28.0000   
SibSp       0.00    8.0000   0.523008   1.102743   0.0000   
Parch       0.00    6.0000   0.381594   0.806057   0.0000   
Fare        0.00  512.3292  32.204208  49.693429  14.4542   
FamilySize  1.00   11.0000   1.904602   1.613459   1.0000   

            variation_coefficient  
Pclass                   0.362149  
Age                      0.443426  
SibSp                    2.108464  
Parch                    2.112344  
Fare                     1.543073  
FamilySize               0.847137

In [372]:
def _convert_fare(X):
    if X < 50:
        return "basic"
    
    elif 50 <= X <= 250:
        return "medium"

    else:
        return "premium"


df["FareCat"] = df["Fare"].apply(_convert_fare).astype("category")
df["FareCat"]

0       basic
1      medium
2       basic
3      medium
4       basic
        ...  
886     basic
887     basic
888     basic
889     basic
890     basic
Name: FareCat, Length: 891, dtype: category
Categories (3, object): ['basic', 'medium', 'premium']

Creamos una columna para saber el nivel en que se encuentra cada pasajero, en caso de null o dato inconsistente completamos con M

In [373]:
df["Deck"] = df["Cabin"].apply(lambda s: s[0] if pd.notnull(s) else 'M')

In [374]:
new_columns = ["Title", "FamilySize", "FareCat", "CabinUnknown", "Deck"]

Ahora evaluaremos las características agregadas

In [375]:
columns = df.columns.tolist()
columns

['Survived',
 'Pclass',
 'Sex',
 'Age',
 'SibSp',
 'Parch',
 'Fare',
 'Cabin',
 'Embarked',
 'Title',
 'FamilySize',
 'CabinUnknown',
 'FareCat',
 'Deck']

In [385]:
original_columns = [
    'Pclass',
    'Sex',
    'Age',
    'SibSp',
    'Parch',
    'Fare',
    'Embarked',
]

feature_columns = [
    'Pclass',
    'Sex',
    'Age',
    'Fare',
    'Embarked',
    'Title',
    'FamilySize',
    'CabinUnknown',
    'FareCat',
    'Deck',
]

columnas_total = list(set(original_columns + feature_columns))

X = df[columnas_total]

y = df["Survived"]

In [386]:
from utils.eda import get_categoric_columns


categoric_columns = get_categoric_columns(df)

for column in columnas_total:
    if column not in categoric_columns:
        continue

    unique_values = df[column].unique().tolist()
    print(f"Valores únicos para columna '{column}': {str(unique_values)}")

Valores únicos para columna 'Sex': ['male', 'female']
Valores únicos para columna 'FareCat': ['basic', 'medium', 'premium']
Valores únicos para columna 'Deck': ['u', 'c', 'e', 'g', 'd', 'a', 'b', 'f', 't']
Valores únicos para columna 'Embarked': ['s', 'c', 'q', 'unk']
Valores únicos para columna 'Title': ['mr', 'mrs', 'miss']


De estos datos, podemos obtener columnas a codificar como ordinales como one hot.

In [387]:
ordinal_features_columns = [
    "FareCat",
]

ordinal_columns_category_feature = [["basic", "medium", "premium"]]

onehot_original_columns = [
    "Sex",
    "Embarked"
]

onehot_feature_columns= [
    "Sex",
    "Title",
    "Deck",
    "Embarked"
]

Importamos las librerias para modelar

In [388]:
from sklearn.compose import ColumnTransformer
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.preprocessing import OrdinalEncoder, OneHotEncoder, MinMaxScaler
from sklearn.pipeline import Pipeline
import xgboost as xgb
from sklearn.metrics import roc_curve, roc_auc_score
from sklearn.model_selection import cross_val_score
from sklearn.metrics import (
    classification_report,
    confusion_matrix,
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
)

In [389]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.3,
    random_state=42,
)
print(X_train.columns)

X_train_original = X_train[original_columns]
X_train_feature = X_train[feature_columns]
X_test_original = X_test[original_columns]
X_test_feature = X_test[feature_columns]

Index(['Age', 'FamilySize', 'Sex', 'FareCat', 'CabinUnknown', 'Deck', 'SibSp',
       'Embarked', 'Title', 'Parch', 'Fare', 'Pclass'],
      dtype='object')


Seteamos preprocesadores

In [394]:
print("Columnas en X_train_original:", X_train_original.columns)

# Preprocesador original
preprocessor_original = ColumnTransformer(
    transformers=[
        ("onehot", OneHotEncoder(), onehot_original_columns)
    ],
    remainder="passthrough",
    force_int_remainder_cols=False
)

# Preprocesador con Feature Engineering
preprocessor_feature_eng = ColumnTransformer(
    transformers=[
        ("ordinal", OrdinalEncoder(), ordinal_features_columns),
        ("onehot", OneHotEncoder(), onehot_feature_columns),
    ],
    remainder="passthrough",
    force_int_remainder_cols=False
)



Columnas en X_train_original: Index(['Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Embarked'], dtype='object')


#### Modelo sin Feature engineering

In [402]:
rf_original = Pipeline(
    steps=[
        ("preprocessor", preprocessor_original),
        ("model", RandomForestClassifier(n_estimators=20, random_state=42))
    ]
)

# Ajustar el modelo

rf_original.fit(X_train_original, y_train)


Pipeline(steps=[('preprocessor',
                 ColumnTransformer(force_int_remainder_cols=False,
                                   remainder='passthrough',
                                   transformers=[('onehot', OneHotEncoder(),
                                                  ['Sex', 'Embarked'])])),
                ('model',
                 RandomForestClassifier(n_estimators=20, random_state=42))])

Evaluamos resultados

In [ ]:
y_pred_rf_original = rf_original.predict(X_test_original)

accuracy_rf_original = accuracy_score(y_test, y_pred_rf_original)
precision_rf_original = precision_score(y_test, y_pred_rf_original)
recall_rf_original = recall_score(y_test, y_pred_rf_original)
f1_rf_original = f1_score(y_test, y_pred_rf_original)
roc_auc_rf_original = roc_auc_score(y_test, y_pred_rf_original)

print(f"Accuracy RF original: {accuracy_rf_original}")
print(f"Precision RF original: {precision_rf_original}")
print(f"Recall RF original: {recall_rf_original}")
print(f"F1 RF original: {f1_rf_original}")
print(f"ROC AUC original: {roc_auc_rf_original}")

Accuracy RF original: 0.7686567164179104
Precision RF original: 0.8064516129032258
Recall RF original: 0.7961783439490446
F1 RF original: 0.8012820512820513
ROC AUC original: 0.7629540368393871


#### Modelo con Feature Engineering

In [400]:
rf_feature_eng = Pipeline(
    steps=[
        ("preprocessor", preprocessor_feature_eng),
        ("model", RandomForestClassifier(n_estimators=20, random_state=42))
    ]
)

rf_feature_eng.fit(X_train_feature, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(force_int_remainder_cols=False,
                                   remainder='passthrough',
                                   transformers=[('ordinal', OrdinalEncoder(),
                                                  ['FareCat']),
                                                 ('onehot', OneHotEncoder(),
                                                  ['Sex', 'Title', 'Deck',
                                                   'Embarked'])])),
                ('model',
                 RandomForestClassifier(n_estimators=20, random_state=42))])

Evaluamos resultados

In [401]:
y_pred_rf_feature_eng = rf_feature_eng.predict(X_test_feature)

accuracy_rf_feature_eng = accuracy_score(y_test, y_pred_rf_feature_eng)
precision_rf_feature_eng = precision_score(y_test, y_pred_rf_feature_eng)
recall_rf_feature_eng = recall_score(y_test, y_pred_rf_feature_eng)
f1_rf_feature_eng = f1_score(y_test, y_pred_rf_feature_eng)
roc_auc_rf_feature_eng = roc_auc_score(y_test, y_pred_rf_feature_eng)


print(f"Accuracy RF Feature Eng: {accuracy_rf_feature_eng}")
print(f"Precision RF Feature Eng: {precision_rf_feature_eng}")
print(f"Recall RF Feature Eng: {recall_rf_feature_eng}")
print(f"F1 RF Feature Eng: {f1_rf_feature_eng}")
print(f"ROC AUC Feature Eng: {roc_auc_rf_feature_eng}")

Accuracy RF Feature Eng: 0.7761194029850746
Precision RF Feature Eng: 0.8129032258064516
Recall RF Feature Eng: 0.802547770700637
F1 RF Feature Eng: 0.8076923076923077
ROC AUC Feature Eng: 0.7706432547196879


Comparamos el rendimiento de ambos modelos

In [404]:
rf_original_metrics = {
    "Accuracy": accuracy_rf_original,
    "Precision": precision_rf_original,
    "Recall": recall_rf_original,
    "F1-Score": f1_rf_original,
    "ROC-AUC": roc_auc_rf_original,
}

rf_feature_eng_metrics = {
    "Accuracy": accuracy_rf_feature_eng,
    "Precision": precision_rf_feature_eng,
    "Recall": recall_rf_feature_eng,
    "F1-Score": f1_rf_feature_eng,
    "ROC-AUC": roc_auc_rf_feature_eng,
}

results = pd.DataFrame(
    [rf_original_metrics, rf_feature_eng_metrics,],
    index=["Random Forests Original", "Random Forests Feature",],)
print(results)

print("\n")

print("Mejores modelos por métrica:\n")
print(results.idxmax(axis=0))

                         Accuracy  Precision    Recall  F1-Score   ROC-AUC
Random Forests Original  0.768657   0.806452  0.796178  0.801282  0.762954
Random Forests Feature   0.776119   0.812903  0.802548  0.807692  0.770643


Mejores modelos por métrica:

Accuracy     Random Forests Feature
Precision    Random Forests Feature
Recall       Random Forests Feature
F1-Score     Random Forests Feature
ROC-AUC      Random Forests Feature
dtype: object


Aplicando Feature engineering se obtienen mejores métricas

## Interpretacion de resultados

Interpretación

Accuracy
- La precisión general del modelo Feature es ligeramente mejor (77.6%) en comparación con el modelo Original (76.9%). 
- Esto significa que el modelo con Feature Engineering clasifica correctamente un porcentaje mayor de muestras.


Precision
- La precisión mide la proporción de predicciones positivas correctas.

- El modelo Feature tiene una precisión del 81.3%, mejorando sobre el 80.6% del modelo original.
- Esto indica que Random Forests Feature es más efectivo para evitar falsos positivos.
- Recall
- El recall mide la proporción de positivos correctamente identificados.

- El modelo Feature obtiene un 80.3%, mientras que el modelo original llega al 79.6%.
- Esto significa que Random Forests Feature identifica más positivos reales.

F1-Score
- La F1-Score es el promedio armónico entre Precision y Recall, siendo una métrica equilibrada.

- El modelo Feature obtiene 80.8%, superando el 80.1% del modelo original.
- Esto refuerza que el modelo con Feature Engineering logra un mejor equilibrio entre Precision y Recall.

ROC-AUC
- El ROC-AUC mide la capacidad del modelo para distinguir entre clases.

- El modelo Feature tiene un 0.7706, ligeramente mejor que el 0.7629 del modelo original.
- Esto indica que Random Forests Feature tiene una mejor capacidad para discriminar entre clases.

Las carácteristicas nuevas ayudaron al modelo a encontrar relaciones de forma más directa, por ejemplo el titulo antes implicito en el nombre ahora se encontraba en un feature especifico, también el nivel de la cabina, el costo del ticket, etc.

Estas caracteristicas hacen que la información se encuentre de forma más explicita para nuestro modelo, cuestiones cómo la cantidad de miembros de la familia puede influir a la hora de saber si una persona tuvo mayor probabilidad de sobrevivir, por ejemplo una familia con muchos hijos buscará buscar a todos los hijos y ayudarlos, una familia solo de marido y mujer se tendrán mutuamente para ayudarse en caso de que sea necesario y no tendrán personas que dependan directamente de ellos y un pasajero sin familia solo buscará la salida sin preocuparse por otros, pero al mismo tiempo no tendrá alguien que lo ayude activamente en caso de que lo necesite.

Las nuevas variables pueden hacer que este tipo de información se observe de forma más explicita